weave-js/src/components/Panel2/PanelGroupingEditor.tsx

```diff
@@ -0,0 +1,227 @@
+import {
+  constFunction,
+  constNodeUnsafe,
+  isFunctionLiteral,
+  varNode,
+  voidNode,
+  Node,
+  constString,
+  opIndex,
+  opPick,
+  pickSuggestions,
+  NodeOrVoidNode,
+  Type,
+  // opLimit,
+} from '@wandb/weave/core';
+import {IconChevronDown} from '@wandb/weave/components/Icon';
+import {MOON_50} from '@wandb/weave/common/css/color.styles';
+
+import * as _ from 'lodash';
+import React, {useCallback, useMemo} from 'react';
+
+import {WeaveExpression} from '../../panel/WeaveExpression';
+import {useMutation, useNodeValue} from '../../react';
+import * as Panel2 from './panel';
+import {PanelContextProvider} from './PanelContext';
+import {ConfigFieldModifiedDropdown, ConfigFieldWrapper} from './ConfigPanel';
+import {VisualEditorMode, getSimpleKeyType} from './visualEditors';
+
+const inputType = {
+  type: 'function' as const,
+  inputTypes: {},
+  outputType: 'any' as const,
+};
+interface PanelGroupingEditorConfig {
+  node: Node;
+}
+
+type PanelGroupingEditorProps = Panel2.PanelProps<
+  typeof inputType,
+  PanelGroupingEditorConfig
+>;
+
+interface VisualGroupingState {
+  key: string;
+}
+
+// GroupingEditor is a generic editor for filters, for any list.
+// But we hardcode a sort order for our monitoring Span type for now.
+// In the future this could be controlled by a config parameter.
+const keySortOrder = (key: string) => {
+  if (key.startsWith('attributes.')) {
+    return 1;
+  }
+  if (key.startsWith('summary.')) {
+    return 2;
+  }
+  if (key === 'id' || key.includes('_id')) {
+    return 4;
+  }
+  return 3;
+};
+
+const groupingExpressionToVisualState = (
+  expr: NodeOrVoidNode
+): VisualGroupingState | null => {
+  if (expr.nodeType === 'void') {
+    return null;
+  }
+  if (expr.nodeType !== 'output') {
+    return null;
+  }
+  if (expr.fromOp.name !== 'pick') {
+    return null;
+  }
+  if (expr.fromOp.inputs.key.nodeType !== 'const') {
+    return null;
+  }
+  return {
+    key: expr.fromOp.inputs.key.val as string,
+  };
+};
+
+const visualGroupingStateToExpression = (
+  groupingState: VisualGroupingState,
+  listItemType: Type
+) => {
+  return opPick({
+    obj: varNode(listItemType, 'row'),
+    key: constString(groupingState.key),
+  });
+};
+
+const setGroupingKey = (
+  groupingState: VisualGroupingState,
```

Comment:
do we need to pass in groupingState here

Reply:
No, but just following a pattern where we add these "action" style functions that modify a groupingState, returning a new one. When groupingState has more stuff in it, it'll make more sense :). But the FilterEditor does and you can see the pattern there.

Reply:
makes sense, the way it was set up i figured we might need it later